src/replication.c

```diff
@@ -2173,6 +2175,208 @@ istgt_lu_resize_volume(spec_t *spec, uint64_t size) {
 	return true;
 }
 
+#define IS_REPLICA_ID_EXIST_IN_LIST(_replica_id, _list, _no_of_items, _is_exist) 	\
+	do {										\
+		int i;									\
+		for (i=0; i < _no_of_items; i++) {					\
+			if (strncmp(_replica_id, _list[i], REPLICA_ID_LEN) == 0) {	\
+					_is_exist = true;				\
+					break;						\
+			}								\
+		}									\
+	} while(0)
+
+
+/* is_valid_known_replica_list this function should be called by taking
+ * spec->rq_mtx lock
+ */
+static bool
+is_valid_known_replica_list(spec_t *spec, int listcnt, char **known_replica_id_list) {
+	trusty_replica_t *trusty_replica = NULL;
+	bool found_replica;
+	int i;
+
+	ASSERT(MTX_LOCKED(&spec->rq_mtx));
+	for(i=0; i<listcnt; i++) {
+		found_replica = false;
+		TAILQ_FOREACH(trusty_replica, &spec->lu->trusty_replicas, next) {
+			if (strncmp(trusty_replica->replica_id,
+			    known_replica_id_list[i], REPLICA_ID_LEN) == 0) {
+				found_replica = true;
+				break;
+			}
+		}
+		// If known replicaid doesn't exist in trusty replica then
+		// replicaid list has wrong information
+		if (!found_replica) {
+			ISTGT_ERRLOG("replicaid (%s) doesn't exist in "
+			    "trusty replica list\n", known_replica_id_list[i]);
+			return false;
+		}
+	}
+	return true;
+}
+
+static bool
+can_replica_remove(spec_t *spec, replica_t *removing_replica) {
+	replica_t *replica;
+
+	if (spec->healthy_rcount < spec->consistency_factor)
+		return false;
+
+	// TODO: Return true when all the replicas_id should be healthy
+	// else return false
+	TAILQ_FOREACH(replica, &spec->rq, r_next) {
+		if (replica != removing_replica &&
+		    replica->state != ZVOL_STATUS_HEALTHY)
+			return false;
+	}
+	return true;
+}
+
+/* istgt_lu_remove_unknown_replica will disconnect the replica and update
+ * volume configurations. Steps performed in this function
+ * 1. Get replica information that is going to remove/disconnect from target
+ * 2. Wait for any Ongoing snapshots
+ * 3. Verify whether replicas are healthy or not other than removing replica. If
+ *    there are degraded replicas return error
+ * 4. On success pause the IOs for 5 seconds. If there are pending IOs return error
+ *    by resuming IOs else continue the process
+ * 5. Mark the replica cordon for IOs
+ * 6. Update inmemory configurations.
+ */
+int
+istgt_lu_remove_unknown_replica(spec_t *spec, int drf, char **known_replica_id_list) {
+	replica_t *replica, *removing_replica = NULL;
+	int rc = -1;
+	// Hard coded IO wait time
+	int io_wait_time = 5;
+	int new_rf = drf;
+	int new_cf = CONSISTENCY_FACTOR(new_rf);
+	int found_in_rq_list = false;
+	int found_replica = false;
+
+	trusty_replica_t *trusty_replica = NULL;
+
+	MTX_LOCK(&spec->rq_mtx);
+	ASSERT(drf == spec->replication_factor - 1);
+
+#ifdef REPLICATION
```

Comment:
Please remove this macro protection.

Reply:
Sorry for not going through changes I will update it